include/gul14/ThreadPool.h

```diff
@@ -0,0 +1,297 @@
+ *
+ * \since GUL version 2.11
+ */
+class ThreadPool
+{
+public:
+    using TaskId = std::uint64_t;
+    using TimePoint = std::chrono::time_point<std::chrono::system_clock>;
+    using Duration = TimePoint::duration;
+
+    /// Maximum possible capacity for the task queue.
+    constexpr static std::size_t max_capacity{ 10'000'000 };
+
+    /// Maximum possible number of threads
+    constexpr static std::size_t max_threads{ 10'000 };
+
+    /**
+     * Create a thread pool with the desired number of threads and the specified capacity
+     * for queuing tasks.
+     *
+     * This constructor launches the desired number of threads. The threads are joined
+     * when the ThreadPool object gets destroyed.
+     *
+     * \param num_threads  Desired number of threads
+     * \param capacity     Maximum number of pending tasks that can be queued
+     *
+     * \exception std::invalid_argument is thrown if the desired number of threads is
+     *            zero or greater than max_threads, or if the requested capacity is zero
+     *            or exceeds max_capacity.
+     */
+    ThreadPool(std::size_t num_threads, std::size_t capacity = 1000);
+
+    /**
+     * Destruct the ThreadPool and join all threads.
+     *
+     * This destructor blocks until all threads have terminated. Work that has not yet
+     * been started in one of the threads gets canceled, but work that has already been
+     * assigned to a thread continues to be executed until it completes.
+     */
+    ~ThreadPool();
+
+    /**
+     * Enqueue a task.
+     *
+     * There are multiple overloads of this function for variations of the arguments:
+     *
+     * \param fct   A function object or function pointer to be executed. This function
+     *              must return void and either takes no arguments (`void fct()`) or
+     *              a reference to the ThreadPool by which it gets executed
+     *              (`void fct(ThreadPool&)`). Null pointers are not allowed.
+     * \param start_time  Earliest time point at which the task is to be started
+     * \param name  Optional name for the task (mainly for debugging)
+     *
+     * \returns a unique ID to identify this task.
+     * \exception std::runtime_error is thrown if the queue is full. std::invalid_argument
+     *            is thrown if a null pointer is given.
+     */
+    TaskId add_task(std::function<void(ThreadPool&)> fct, TimePoint start_time = {},
+        std::string name = {});
+
+    TaskId add_task(std::function<void()> fct, TimePoint start_time = {},
+        std::string name = {});
+
+    /**
+     * \overload
+     *
+     * \param fct   A function object or function pointer to be executed. This function
+     *              must return void and either takes no arguments (`void fct()`) or
+     *              a reference to the ThreadPool by which it gets executed
+     *              (`void fct(ThreadPool&)`). Null pointers are not allowed.
```

Comment:
Maybe mention that the reference can dangle later on (i.e. when the `ThreadPool` is destructed and our task is still running. `ThreadPool` blocks in it's dtor, but its lifetime already ended when the dtor has been started.

We might got a signal (at least I saw some `condition_variable` above, but the task must not call anything in the `ThreadPool` at this time.

We should at least warn people. This seems not too far fetched.

I also do not know what the `cv` triggers in reality, but that already can be problematic.

Reply:
The lifetime of the `ThreadPool` object formally ends when the destructor call starts, but that does not mean that the reference already dangles: As long as the destructor body is still being executed, a partially-functional object "ruin" still exists – otherwise the destructor could not do its work. What "partially-functional" means is defined by what the destructor has already done. 

It is a bit sketchy – but I would say that in our case all public member functions are still safe to use until the destructor has finished its work. And once that is done, there are no more threads whose references could dangle.

The `cv_` triggers when
- a task has been added or
- thread pool shutdown has been requested (e.g. by the destructor).

Reply:
According to the standard calls from outside the destructed object are forbidden at the moment the destructor is entered.
What you mean is inside the to-be-destructed object.

And well, look at the code with the safe_self_, maybe you like that and it is 100% safe and easy to grasp for library users.

Reply:
> According to the standard calls from outside the destructed object are forbidden at the moment the destructor is entered.

I do not see that. According to https://stackoverflow.com/questions/46181083/is-it-safe-to-wait-for-asynchronous-work-by-joining-in-the-destructor , [class.cdtor]/1 says

> For an object with a non-trivial destructor, referring to any non-static member or base class of the object _after the destructor finishes_ execution results in undefined behavior.

Notably, this talks about *after the destructor finishes*, not about when it starts, and it cannot work any other way. When the destructor starts, the object at its level of "constructedness" must necessarily still exist. In a hierarchy where A is base of B and B is base of C, inside `~B()` it must be guaranteed that
- members of A are still present
- members of B are still present as far as ~B() has not done anything to them yet.
All members of C are guaranteed to have been cleaned up by ~C() at this point.

If that is true, the rest is defense against data races, and I think we are safe in that regard (but we should have another look).

So, the "safe self" code is fine, but I do not think we need it in this case.

Reply:
https://en.cppreference.com/w/cpp/language/lifetime#Access_outside_of_lifetime

<img width="965" alt="Screenshot 2024-02-03 at 09 59 44" src="https://github.com/gul-cpp/gul14/assets/16012374/7290f159-3c02-4479-83b8-c52793d7384b">

Ok.

Can you give a use-case for the `ThreadPool` reference?

I can imagine it is intended for tasks to add another talks themselves.

But then we run into a deadlock.
The destructor of `ThreadPool` takes the lock.
The task (`fct()`) reached a point where it wants to start its subtask or continuation.
It calls `ThreadPool.add_task()`.
Adding the task tries to take the lock.

The task waits for the lock.
The lock will only be released when the task finishes.

<img width="1006" alt="Screenshot 2024-02-03 at 10 05 11" src="https://github.com/gul-cpp/gul14/assets/16012374/edd2b202-2406-4ff8-882f-02625f490848">

_The `task_id` closure is immediately executed_

I guess `add_task()` (and possibly other member functions as well) should look at `cancel_requested_`, before trying to take the lock (but instead return immediately or something).
But even that is strictly speaking not enough, because the dtor() takes the lock first and then sets the cancel variable. That would need to be swapped around (first mark we-want-to-destruct, then lock).
And `cancel_requested_` needs to be some thread safe thing then.

I believe it is far easier to hand out weak_ptr because we do not need to think hard about all the possible races and so on, it just magically works as expected. Dont be smart, just keep stuff simple.

Reply:
> Can you give a use-case for the `ThreadPool` reference?
> 
> I can imagine it is intended for tasks to add another talks themselves.

Exactly. The idea was for a task to schedule its own continuation. For instance, we often have tasks like "do A, wait 10 milliseconds, do B". With the `ThreadPool` reference, task A can simply do its thing and schedule task B to start 10 ms later, freeing the thread for other work in the meantime.

We would really want coroutines for that kind of thing, but... well, we can still dream. ;)

> But then we run into a deadlock. The destructor of `ThreadPool` takes the lock. The task (`fct()`) reached a point where it wants to start its subtask or continuation. It calls `ThreadPool.add_task()`. Adding the task tries to take the lock.

Wait... I fail to see the deadlock. The `ThreadPool` destructor takes the lock only to set `cancel_requested` and then immediately releases it. The worker threads will not start a new task once that is set. Am I missing some subtlety?

> I believe it is far easier to hand out weak_ptr because we do not need to think hard about all the possible races and so on, it just magically works as expected. Dont be smart, just keep stuff simple.

It seems to me that this is where we disagree. I still do not see how this smart pointer juggling makes anything easier in this context. I can see a use case for it if we hand out more complex objects instead of `ThreadId`s, but not here.

Reply:
> Wait... I fail to see the deadlock.

> The ThreadPool destructor takes the lock only to set cancel_requested and then immediately releases it.

I see this code
```c++
ThreadPool::~ThreadPool()
{
    std::unique_lock<std::mutex> lock(mutex_);
    cancel_requested_ = true;
    lock.unlock();
    cv_.notify_all();

    for (auto& t : threads_)
    {
        if (t.joinable())
            t.join();
    }
}
```

* Take the lock
* Set `cancel_request_`
* :woman_facepalming: 

Ok, you are right, I never release a lock like that, so I misread it. It took several reads for me to notice that.
(If I use a lock like that I always release it be leaving the scope and never by the member function. I would throw an additional scope just around the function potion that should be locked.)

What I would have expected was something like this

```c++
ThreadPool::~ThreadPool()
{
    {
        std::unique_lock<std::mutex> lock(mutex_);
        cancel_requested_ = true;
    }
    cv_.notify_all();

    for (auto& t : threads_)
    {
        if (t.joinable())
            t.join();
    }
}
```

I use that style (and expected it, my bad) because it makes it impossible to forget the release. Calling `unlock` smells like handling a pure mutex and not a lock. :woman_shrugging: But thats probably purely personal.

https://en.cppreference.com/w/cpp/thread/lock
See the examples, just as I always use it.

Anyhow, it seems all good now, just did not read careful enough. Maybe change style but otoh I do not know how common the scope style is?!

And I still believe `add_task` needs to check for `cancel_requested_` and silently (?) drop requests for more tasks if they will never be scheduled anyhow?

----

> It seems to me that this is where we disagree. I still do not see how this smart pointer juggling makes anything easier in this context. I can see a use case for it if we hand out more complex objects instead of `ThreadId`s, but not here.

Well, I believe we should not hand out ~~filehandle integer numbers~~ dump IDs that people would have to feed back into a (the correct one to be honest) `ThreadPool` object; but instead

```
tp = ThreadPool{ };
 ...
a_task = tp.add_task(...);
 ...

// Is the task running? (version 1)
tp.is_running(a_task);

// Is the task running? (object-y version)
a_task.is_running();

and of course here `tp` might live, but maybe `a_task` outlived `tp`.

I can not see why you are so 'repelled' by 'smart pointer juggling'. I believe that makes all the APIs much easier.

But with one point you are right, if it was only `foo(tp)` that takes the ThreadPool by reference, that in itself is no problem with the current teardown behavior. It still can come up later when we would allow for example detatched threads or the ThreadPool to die before all its tasks die. But ... I guess without a real `TaskID` replacement object it is ok as is.

Reply:
> a_task = tp.add_task(...);
> ...
>// Is the task running? (object-y version)
>a_task.is_running();

This is now implemented. I have removed the reverse call `pool.is_running(task)` from the API just to keep things less cluttered.